File: PID.py
class PID:
    def __init__(self, Kp, Ki, Kd, output_limits=(None, None)):
        """
        PID controller

        Args:
            Kp, Ki, Kd : floats, PID gains
            output_limits : (min, max), saturation limits for control output
        """
        self.Kp = Kp
        self.Ki = Ki
        self.Kd = Kd
        self.min_output, self.max_output = output_limits

        self.integral = 0.0
        self.prev_error = 0.0
        self.first_call = True

    def update(self, setpoint, measurement, dt):
        """
        Compute PID output

        Args:
            setpoint : desired target value
            measurement : current measured value
            dt : timestep

        Returns:
            control_output : float
        """
        error = setpoint - measurement

        # Proportional
        P = self.Kp * error

        # Integral
        self.integral += error * dt
        I = self.Ki * self.integral

        # Derivative (guard against first call)
        if self.first_call:
            D = 0.0
            self.first_call = False
        else:
            derivative = (error - self.prev_error) / dt
            D = self.Kd * derivative

        # Save error
        self.prev_error = error

        # PID sum
        output = P + I + D

        # Apply output limits
        if self.min_output is not None:
            output = max(self.min_output, output)
        if self.max_output is not None:
            output = min(self.max_output, output)

        return output

File: controller_in_rust/src/pid.rs
pub struct Pid {
    kp: f64,
    ki: f64,
    kd: f64,
    min_output: f64,
    max_output: f64,
    integral: f64,
    prev_error: f64,
    first_call: bool,
}

impl Pid {
    pub fn new(kp: f64, ki: f64, kd: f64, min_output: f64, max_output: f64) -> Self {
        Pid {
            kp,
            ki,
            kd,
            min_output,
            max_output,
            integral: 0.0,
            prev_error: 0.0,
            first_call: true,
        }
    }

    pub fn update(&mut self, setpoint: f64, measurement: f64, dt: f64) -> f64 {
        let error = setpoint - measurement;

        let p = self.kp * error;

        self.integral += error * dt;
        let i = self.ki * self.integral;

        let d = if self.first_call {
            self.first_call = false;
            0.0
        } else {
            let derivative = (error - self.prev_error) / dt;
            self.kd * derivative
        };

        self.prev_error = error;

        (p + i + d).clamp(self.min_output, self.max_output)
    }
}
